Improve `EditorDirDialog`

timothyqiu opened the pull request:
* Automatically selects the newly created directory
* Automatically selects "res://" when nothing is selected
* Fixes an error when overwrite/replace dialog appears
* Changes "copy checkbox + action button" to "copy button + move button"
* Double clicking a directory (un)collapses it instead of copy/move
* Moves "Create Folder" to top row for better UX
* Uses DirectoryCreateDialog for "Create Folder"

| Before | After |
|---|---|
| ![ksnip_20231224-140532](https://github.com/godotengine/godot/assets/372476/3a7d559c-cb73-4cf6-85e1-6bda236e5303) | ![image](https://github.com/godotengine/godot/assets/372476/42e0509a-0dbb-4594-9878-37d607cdbeaa) |

timothyqiu committed:
Improve EditorDirDialog

* Automatically selects the newly created directory
* Automatically selects "res://" when nothing is selected
* Fixes an error when overwrite/replace dialog appears
* Changes "copy checkbox + action button" to "copy button + move button"
* Double clicking a directory (un)collapses it instead of copy/move
* Uses DirectoryCreateDialog for "Create Folder"

KoBeWi commented on editor/gui/editor_dir_dialog.cpp:
```diff
@@ -134,29 +129,35 @@ void EditorDirDialog::_item_collapsed(Object *p_item) {
+	bool disable = tree->get_selected() == nullptr;
+	get_ok_button()->set_disabled(disable);
+	copy->set_disabled(disable);
+	makedir->set_disabled(disable);
```
This one could default to `res://` if nothing is selected.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1883,8 +1883,8 @@ Vector<String> FileSystemDock::_check_existing() {
-void FileSystemDock::_move_dialog_confirm(const String &p_path) {
-	_move_operation_confirm(p_path, move_dialog->is_copy_pressed());
+void FileSystemDock::_move_dialog_confirm(const String &p_path, bool p_copy) {
+	_move_operation_confirm(p_path, p_copy);
```
You could connect to this method directly. The intermediate one was only needed for getting the bool.

timothyqiu replied:
The `connect()` does not support default argument. In order to connect to `_move_operation_confirm` directly, I have to bind `OVERRITE_UNDECIDED`. But in order to bind it, it seems I have to register the enum for binding. Since the class and enum is not exposed for scripting, I think doing this makes things more complicated.

timothyqiu replied:
OK, I did the change. Note that `Overwrite` enum is now changed to public and bind.

KoBeWi commented on editor/gui/editor_dir_dialog.h:
```diff
@@ -46,20 +46,22 @@ class EditorDirDialog : public ConfirmationDialog {
 
 	Tree *tree = nullptr;
 	bool updating = false;
-	CheckBox *copy = nullptr;
```
You should also remove CheckBox forward declaration.

KoBeWi commented on editor/gui/editor_dir_dialog.cpp:
```diff
@@ -185,13 +186,15 @@ void EditorDirDialog::_make_dir_confirm() {
 
 void EditorDirDialog::_bind_methods() {
-	ADD_SIGNAL(MethodInfo("dir_selected", PropertyInfo(Variant::STRING, "dir")));
+	ADD_SIGNAL(MethodInfo("copy_selected", PropertyInfo(Variant::STRING, "dir")));
+	ADD_SIGNAL(MethodInfo("move_selected", PropertyInfo(Variant::STRING, "dir")));
```
```suggestion
	ADD_SIGNAL(MethodInfo("copy_pressed", PropertyInfo(Variant::STRING, "dir")));
	ADD_SIGNAL(MethodInfo("move_pressed", PropertyInfo(Variant::STRING, "dir")));
```

KoBeWi commented on editor/gui/editor_dir_dialog.cpp:
```diff
@@ -204,14 +207,14 @@ EditorDirDialog::EditorDirDialog() {
 	tree = memnew(Tree);
 	vb->add_child(tree);
 	tree->set_v_size_flags(Control::SIZE_EXPAND_FILL);
-	tree->connect("item_activated", callable_mp(this, &EditorDirDialog::_item_activated));
```
This makes it impossible to accept with double click. Maybe it could default to Move? (and could copy if you hold Ctrl, idk)

timothyqiu replied:
I changed double clicking to (un)collapse that directory. Consistent with the filesystem dock.

Previously, double clicking moves/copies depending on the last action performed (because "Copy Files(s)" state is kept between runs).

zolrath commented:
This may not be important but in every other dialog and menu I can see in Godot, Cancel is the last/far right option.
However, in this menu, Cancel is a middle option.

Is there a UI/UX standard for the location of the Cancel button in Godot?

timothyqiu commented:
Moved "Create Folder" to the top row so that the Cancel button is no longer in the middle :P 

![image](https://github.com/godotengine/godot/assets/372476/42e0509a-0dbb-4594-9878-37d607cdbeaa)

Also changed "Create Folder" to use `DirectoryCreateDialog`, the same as FileSystem dock.

![image](https://github.com/godotengine/godot/assets/372476/453de4d3-74a0-458d-b1f6-fc4f0465c689)

KoBeWi commented on editor/gui/editor_dir_dialog.cpp:
```diff
@@ -59,13 +55,35 @@ void EditorDirDialog::_update_dir(TreeItem *p_item, EditorFileSystemDirectory *p
+		set_title(vformat(TTR("Move/Duplicate: %s"), path.get_file()));
+	} else {
+		// TRANSLATORS: %d is the number of files that will be moved or duplicated.
+		set_title(vformat(TTRN("Move/Duplicate %d Item", "Move/Duplicate %d Items", p_paths.size()), p_paths.size()));
```
The single item string will be unused 🤔

timothyqiu replied:
It's necessary in order to use `TTRN`. Using `TTRN` here for languages that have more complex plural forms.

KoBeWi replied:
Yep, I get why TTRN is here. I wonder if there is a way to avoid useless string.

KoBeWi replied:
Actually nevermind 🤦‍♂️

KoBeWi commented on editor/gui/editor_dir_dialog.cpp:
```diff
@@ -59,13 +55,35 @@ void EditorDirDialog::_update_dir(TreeItem *p_item, EditorFileSystemDirectory *p
 		p_item->set_text(0, p_dir->get_name());
 	}
 
+	if (path == new_dir_path) {
+		p_item->select(0);
+	} else if (!p_item->get_parent()) {
```
```suggestion
	if (path == new_dir_path || !p_item->get_parent()) {
```

akien-mga commented:
Thanks!
